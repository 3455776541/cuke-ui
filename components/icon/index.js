import React from "react";
import {
	FiVolume2,
	FiLoader,
	FiCheckCircle,
	FiXCircle,
	FiAlertCircle,
	FiX,
	FiChevronUp,
	FiArrowDown
} from "react-icons/fi";

export const InfoIcon = props => <FiVolume2 {...props} />;
export const LoadingIcon = props => <FiLoader {...props} />;
export const SuccessIcon = props => <FiCheckCircle {...props} />;
export const ErrorIcon = props => <FiXCircle {...props} />;
export const WarningIcon = props => <FiAlertCircle {...props} />;
export const CloseIcon = props => <FiX {...props} />;
export const UpIcon = props => <FiChevronUp {...props} />;
export const DownIcon = props => <FiArrowDown {...props} />;
