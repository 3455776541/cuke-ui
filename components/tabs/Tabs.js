import React, { PureComponent, cloneElement } from "react";
import cls from "classnames";
import PropTypes from "prop-types";

const cardType = "card";

export default class Tabs extends PureComponent {
	state = {
		actionKey: "1",
		lineWidth: 0,
		lineOffsetLeft: 0
	};
	static defaultProps = {
		prefixCls: "cuke-tabs",
		defaultActiveKey: "1",
		onChange: () => {}
	};

	static propTypes = {
		prefixCls: PropTypes.string.isRequired,
		defaultActiveKey: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
		type: PropTypes.oneOf([cardType]),
		tabBarExtraContent: PropTypes.oneOfType([
			PropTypes.string,
			PropTypes.number,
			PropTypes.object
		]),
		onChange: PropTypes.func
	};
	componentWillMount() {
		const { defaultActiveKey } = this.props;
		this.setState({
			actionKey: ~~defaultActiveKey
		});
	}
	componentDidMount() {
		if (this.props.type !== cardType) {
			setTimeout(() => {
				this.setActiveLineStyle();
			}, 0);
		}
	}

	setActiveLineStyle = () => {
		const activeElement = this.activeTab;
		const { width, left } = activeElement.getBoundingClientRect();
		const { left: headerOffset } = this.tabsHeader.getBoundingClientRect();
		this.setState({
			lineWidth: width,
			lineOffsetLeft: left - headerOffset
		});
	};
	onTabChange = key => {
		this.setState({ actionKey: key }, () => {
			if (this.props.type !== cardType) {
				this.setActiveLineStyle();
			}
		});
		this.props.onChange(key);
	};

	render() {
		const {
			prefixCls,
			className,
			type,
			tabBarExtraContent,
			children,
			...attr
		} = this.props;

		const { actionKey, lineWidth, lineOffsetLeft } = this.state;

		const content = React.Children.map(children, (element, index) => {
			const key = (index + 1) >> 0;
			return cloneElement(element, {
				actionKey,
				visible: actionKey === key,
				key: index
			});
		});

		const header = React.Children.map(
			children,
			({ props: { tab, disabled } }, index) => {
				const key = (index + 1) >> 0;
				const bindActiveRef =
					actionKey === key ? { ref: node => (this[`activeTab`] = node) } : {};
				return (
					<div
						key={index}
						role="tab"
						aria-disabled={false}
						aria-selected={true}
						className={cls(`${prefixCls}-tab`, {
							[`${prefixCls}-tab-active`]: actionKey === key,
							[`${prefixCls}-tab-disabled`]: !!disabled
						})}
						{...bindActiveRef}
						onClick={() => !disabled && this.onTabChange(key)}
					>
						{tab}
					</div>
				);
			}
		);

		const isCardType = type === cardType;
		return (
			<div className={cls(prefixCls, className)} {...attr}>
				<div
					className={cls(`${prefixCls}-header`, {
						[`${prefixCls}-card`]: isCardType
					})}
					ref={node => (this.tabsHeader = node)}
				>
					{header}
					{!isCardType ? (
						<div
							className={cls(`${prefixCls}-line`)}
							style={{
								width: lineWidth,
								transform: `translate3d(${lineOffsetLeft}px,0,0)`
							}}
						/>
					) : (
						undefined
					)}
					{tabBarExtraContent ? (
						<div className={cls(`${prefixCls}-extra`)}>
							{tabBarExtraContent}
						</div>
					) : (
						undefined
					)}
				</div>
				<div className={cls(`${prefixCls}-content`)}>{content}</div>
			</div>
		);
	}
}
