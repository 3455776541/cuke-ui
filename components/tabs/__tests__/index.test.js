import assert from "power-assert";
import React from "react";
import { render, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Tabs from "../index";

import { SuccessIcon, ErrorIcon } from "../../icon";

describe("<Tabs/>", () => {
  it("should render Tabs", () => {
    const wrapper = render(
      <Tabs defaultActiveKey="1">
        <Tabs.TabPane tab="选项1" key="1">
          1
        </Tabs.TabPane>
        <Tabs.TabPane tab="选项2" key="2">
          2
        </Tabs.TabPane>
        <Tabs.TabPane tab="选项3" key="3">
          3
        </Tabs.TabPane>
      </Tabs>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render custom active key", () => {
    const wrapper = render(
      <Tabs defaultActiveKey="2">
        <Tabs.TabPane tab="选项1" key="1">
          1
        </Tabs.TabPane>
        <Tabs.TabPane tab="选项2" key="2">
          2
        </Tabs.TabPane>
        <Tabs.TabPane tab="选项3" key="3">
          3
        </Tabs.TabPane>
      </Tabs>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render custom tab title", () => {
    const wrapper = shallow(
      <Tabs defaultActiveKey="2">
        <Tabs.TabPane tab={<SuccessIcon />} key="1">
          1
        </Tabs.TabPane>
        <Tabs.TabPane tab={<ErrorIcon />} key="2">
          2
        </Tabs.TabPane>
      </Tabs>
    );
    assert(wrapper.find(SuccessIcon).length === 1);
    assert(wrapper.find(ErrorIcon).length === 1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should find cuke-tabs classnames", () => {
    const wrapper = shallow(
      <Tabs defaultActiveKey="2">
        <Tabs.TabPane tab="选项1" key="1">
          1
        </Tabs.TabPane>
        <Tabs.TabPane tab="选项2" key="2">
          2
        </Tabs.TabPane>
        <Tabs.TabPane tab="选项3" key="3">
          3
        </Tabs.TabPane>
      </Tabs>
    );
    assert(wrapper.find(".cuke-tabs").length === 1);
    assert(wrapper.find(".cuke-tabs-header").length === 1);
    assert(wrapper.find(".cuke-tabs-content").length === 1);
    assert(wrapper.find(".cuke-tabs-line").length === 1);
  });

  it("should render disabled tab", () => {
    const wrapper = render(
      <Tabs defaultActiveKey="2">
        <Tabs.TabPane tab="选项1" key="1">
          1
        </Tabs.TabPane>
        <Tabs.TabPane tab="选项2" disabled key="2">
          2
        </Tabs.TabPane>
        <Tabs.TabPane tab="选项3" key="3">
          3
        </Tabs.TabPane>
      </Tabs>
    );
    assert(wrapper.find(".cuke-tabs-tab-disabled").length === 1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should can trigger onChange event", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Tabs defaultActiveKey="2" onChange={onChange}>
        <Tabs.TabPane tab="选项1" key="1">
          1
        </Tabs.TabPane>
        <Tabs.TabPane tab="选项2" disabled key="2">
          2
        </Tabs.TabPane>
        <Tabs.TabPane tab="选项3" key="3">
          3
        </Tabs.TabPane>
      </Tabs>
    );
    setTimeout(() => {
      wrapper.find(".cuke-tabs-tab").simulate("click");
      expect(onChange).toHaveBeenCalled();
    });
  });
});
