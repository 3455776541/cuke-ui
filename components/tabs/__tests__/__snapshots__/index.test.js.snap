// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs/> should render Tabs 1`] = `
<div
  class="cuke-tabs"
  defaultactivekey="1"
>
  <div
    class="cuke-tabs-header"
  >
    <div
      aria-disabled="false"
      aria-selected="true"
      class="cuke-tabs-tab cuke-tabs-tab-active"
      role="tab"
    >
      选项1
    </div>
    <div
      aria-disabled="false"
      aria-selected="true"
      class="cuke-tabs-tab"
      role="tab"
    >
      选项2
    </div>
    <div
      aria-disabled="false"
      aria-selected="true"
      class="cuke-tabs-tab"
      role="tab"
    >
      选项3
    </div>
    <div
      class="cuke-tabs-line"
      style="width:0;transform:translate3d(0px,0,0)"
    />
  </div>
  <div
    class="cuke-tabs-content"
  >
    <div
      actionkey="1"
      class="cuke-tabs-panel"
      tab="选项1"
    >
      1
    </div>
  </div>
</div>
`;

exports[`<Tabs/> should render custom active key 1`] = `
<div
  class="cuke-tabs"
  defaultactivekey="2"
>
  <div
    class="cuke-tabs-header"
  >
    <div
      aria-disabled="false"
      aria-selected="true"
      class="cuke-tabs-tab"
      role="tab"
    >
      选项1
    </div>
    <div
      aria-disabled="false"
      aria-selected="true"
      class="cuke-tabs-tab cuke-tabs-tab-active"
      role="tab"
    >
      选项2
    </div>
    <div
      aria-disabled="false"
      aria-selected="true"
      class="cuke-tabs-tab"
      role="tab"
    >
      选项3
    </div>
    <div
      class="cuke-tabs-line"
      style="width:0;transform:translate3d(0px,0,0)"
    />
  </div>
  <div
    class="cuke-tabs-content"
  >
    <div
      actionkey="2"
      class="cuke-tabs-panel"
      tab="选项2"
    >
      2
    </div>
  </div>
</div>
`;

exports[`<Tabs/> should render custom tab title 1`] = `
<div
  className="cuke-tabs"
  defaultActiveKey="2"
  onChange={[Function]}
>
  <div
    className="cuke-tabs-header"
  >
    <div
      aria-disabled={false}
      aria-selected={true}
      className="cuke-tabs-tab"
      key="0/.$1"
      onClick={[Function]}
      role="tab"
    >
      <SuccessIcon />
    </div>
    <div
      aria-disabled={false}
      aria-selected={true}
      className="cuke-tabs-tab cuke-tabs-tab-active"
      key="1/.$2"
      onClick={[Function]}
      role="tab"
    >
      <ErrorIcon />
    </div>
    <div
      className="cuke-tabs-line"
      style={
        Object {
          "transform": "translate3d(0px,0,0)",
          "width": 0,
        }
      }
    />
  </div>
  <div
    className="cuke-tabs-content"
  >
    <BreadcrumbItem
      actionKey={2}
      key="0/.$1"
      prefixCls="cuke-tabs-panel"
      tab={<SuccessIcon />}
      visible={false}
    >
      1
    </BreadcrumbItem>
    <BreadcrumbItem
      actionKey={2}
      key="1/.$2"
      prefixCls="cuke-tabs-panel"
      tab={<ErrorIcon />}
      visible={true}
    >
      2
    </BreadcrumbItem>
  </div>
</div>
`;

exports[`<Tabs/> should render disabled tab 1`] = `
<Tabs
  defaultActiveKey="2"
  onChange={[Function]}
  prefixCls="cuke-tabs"
>
  <div
    className="cuke-tabs"
    defaultActiveKey="2"
    onChange={[Function]}
  >
    <div
      className="cuke-tabs-header"
    >
      <div
        aria-disabled={false}
        aria-selected={true}
        className="cuke-tabs-tab"
        key="0/.$1"
        onClick={[Function]}
        role="tab"
      >
        选项1
      </div>
      <div
        aria-disabled={false}
        aria-selected={true}
        className="cuke-tabs-tab cuke-tabs-tab-active cuke-tabs-tab-disabled"
        key="1/.$2"
        onClick={[Function]}
        role="tab"
      >
        选项2
      </div>
      <div
        aria-disabled={false}
        aria-selected={true}
        className="cuke-tabs-tab"
        key="2/.$3"
        onClick={[Function]}
        role="tab"
      >
        选项3
      </div>
      <div
        className="cuke-tabs-line"
        style={
          Object {
            "transform": "translate3d(0px,0,0)",
            "width": 0,
          }
        }
      />
    </div>
    <div
      className="cuke-tabs-content"
    >
      <BreadcrumbItem
        actionKey={2}
        key="0/.$1"
        prefixCls="cuke-tabs-panel"
        tab="选项1"
        visible={false}
      />
      <BreadcrumbItem
        actionKey={2}
        disabled={true}
        key="1/.$2"
        prefixCls="cuke-tabs-panel"
        tab="选项2"
        visible={true}
      >
        <div
          actionKey={2}
          className="cuke-tabs-panel"
          disabled={true}
          tab="选项2"
        >
          2
        </div>
      </BreadcrumbItem>
      <BreadcrumbItem
        actionKey={2}
        key="2/.$3"
        prefixCls="cuke-tabs-panel"
        tab="选项3"
        visible={false}
      />
    </div>
  </div>
</Tabs>
`;
